### PR TITLE
feat: Add DIDComm Shorten URL 1.0 module

### PR DESCRIPTION
**Pull request by gabrielmatau79**
## Summary
This PR adds a new package implementing the DIDComm **Shorten URL 1.0** protocol for @credo-ts/core agents. We need a reusable, well-typed module for DIDComm Shorten URL to support the mediator and related 2060.io agents.

Spec: https://didcomm.org/shorten-url/1.0/

- Inbound message handlers + service that emits typed events
- Public API to request ,return, invalidate shortened URLs
- Tests
- Documentation about how to use.

### CODE REVIEW

File: packages/shorten-url/src/DidCommShortenUrlApi.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      expiresTime: options.expiresTime,
+    })
+
+    await this.messageSender.sendMessage(
+      new OutboundMessageContext(message, { agentContext: this.agentContext, connection }),
+    )
+    return { messageId: message.id }
+  }
+
+  public async invalidateShortenedUrl(options: { connectionId: string; shortenedUrl: string; threadId?: string }) {
```
Comment (genaris) [inline packages/shorten-url/src/DidCommShortenUrlApi.ts]: It looks like the invalidate does not use threadId at all, since according to the protocol it seems that shortenedUrl would be used to identify the flow.

File: packages/shorten-url/src/DidCommShortenUrlEvents.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import type { BaseEvent } from '@credo-ts/core'
+
+export enum ShortenUrlEventTypes {
+  RequestShortenedUrlReceived = 'RequestShortenedUrlReceived',
+  ShortenedUrlReceived = 'ShortenedUrlReceived',
+  InvalidateShortenedUrlReceived = 'InvalidateShortenedUrlReceived',
+}
```
Comment (genaris) [inline packages/shorten-url/src/DidCommShortenUrlEvents.ts]:
```suggestion
export enum DidCommShortenUrlEventTypes {
  DidCommRequestShortenedUrlReceived = 'DidCommRequestShortenedUrlReceived',
  DidCommShortenedUrlReceived = 'DidCommShortenedUrlReceived',
  DidCommInvalidateShortenedUrlReceived = 'DidCommInvalidateShortenedUrlReceived',
}
```

Lately we are prefixing interfaces and events with "DIDComm"as much as we can to distinguish from other scopes like OpenID4VC.

File: packages/shorten-url/src/DidCommShortenUrlService.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  public createShortenedUrl(options: { threadId: string; shortenedUrl: string; expiresTime?: number }) {
+    return new ShortenedUrlMessage(options)
+  }
+
+  public createInvalidate(options: { shortenedUrl: string; threadId?: string }) {
+    return new InvalidateShortenedUrlMessage(options)
+  }
+
+  // Process inbound messages and emit events
+  public async processRequest(ctx: InboundMessageContext<RequestShortenedUrlMessage>) {
```
Comment (genaris) [inline packages/shorten-url/src/DidCommShortenUrlService.ts]:
```suggestion
  public async processRequest(inboundMessageContext: InboundMessageContext<RequestShortenedUrlMessage>) {
```

To follow usual patterns we have in credo-ts

**Review by genaris — CHANGES_REQUESTED** [review]
This is an excellent initial approach! I like the usage of DidCommShortenUrlEvents to not make this module too restrictive and let the application decide how to manage the lifecycle.

But I'm thinking that it would be useful for this module to store specific records in the wallet (e.g. `DidCommShortenUrlRecord`) like other DIDComm protocol extension modules do (look at action-menu and question-answer in credo-ts repo), not only to make the life simpler to the app using this (such as our mediator, which currently do not have any other DB than the one used for Credo wallet), but also to do some important security checks, such as the connection requesting URL invalidation is the one which actually requested to create it.

Another benefit for the records is that they can store the role, state and other data and can be queried for cleaning up if our application needs to do that in the future.

File: packages/shorten-url/src/DidCommShortenUrlApi.ts (hunk 1 above)
Comment (gabrielmatau79) [reply packages/shorten-url/src/DidCommShortenUrlApi.ts]: Resolved in the commit [9121d8b](https://github.com/2060-io/credo-ts-didcomm-ext/pull/109/commits/9121d8b3718462e72d9c94b33577625187928e3c)

File: packages/shorten-url/src/DidCommShortenUrlApi.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+      expiresTime: options.expiresTime,
+    })
+    // Update or create record
+    const record = await this.shortenUrlRepository.findSingleByQuery(this.agentContext, {
+      connectionId: connection.id,
+      requestId: options.threadId,
+      role: ShortenUrlRole.UrlShortener,
+    })
+    // Update existing record
+    if (record) {
```
Comment (genaris) [inline packages/shorten-url/src/DidCommShortenUrlApi.ts]: Shouldn't we check the state of the record to see whether we've already created a short URL, and in such case throw an error? I don't see how a recipient will handle the case where it receives two different short URLs for the same flow (I think it is an error to do so).

File: packages/shorten-url/src/DidCommShortenUrlApi.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      connectionId: connection.id,
+      shortenedUrl: options.shortenedUrl,
+      role: ShortenUrlRole.LongUrlProvider,
+    })
+
+    if (!record) {
+      throw new CredoError('No shorten-url record found for the provided shortened_url on this connection')
+    }
+
+    record.state = ShortenUrlState.InvalidationSent
```
Comment (genaris) [inline packages/shorten-url/src/DidCommShortenUrlApi.ts]: We can also check for current status of this record to see if we've already invalidated it.

**Review by genaris — CHANGES_REQUESTED** [review]
I think in general it's great and we can start releasing an initial version to test with our mediator and Hologram.  

I left some comments, nothing really extraordinary but I just want to check some edge cases (I honestly did not read the spec entirely so please let me know if I'm wrong... or the spec is not right!).

File: packages/shorten-url/src/DidCommShortenUrlService.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+    const requestId = inboundMessageContext.message.id
+
+    const existingRecord = await this.repository.findSingleByQuery(inboundMessageContext.agentContext, {
+      connectionId: connection.id,
+      requestId,
+      role: ShortenUrlRole.UrlShortener,
+    })
+
+    if (existingRecord) {
```
[the remark this answers is not in the document]
Comment (gabrielmatau79) [reply packages/shorten-url/src/DidCommShortenUrlService.ts]: Done! resolved in the commit [7dbef40](https://github.com/2060-io/credo-ts-didcomm-ext/pull/109/commits/7dbef40a8ad5afa8b9bf448552e7c0c80d545e0e)

File: packages/shorten-url/src/repository/DidCommShortenUrlRecord.ts [hunk 7]
```diff
@@ -56,7 +56,7 @@ export class DidCommShortenUrlRecord extends BaseRecord {
       connectionId: this.connectionId,
       role: this.role,
       state: this.state,
-      ...(this.requestId ? { requestId: this.requestId } : {}),
+      ...(this.threadId ? { threadId: this.threadId } : {}),
       ...(this.shortenedUrl ? { shortenedUrl: this.shortenedUrl } : {}),
```
Comment (genaris) [inline packages/shorten-url/src/repository/DidCommShortenUrlRecord.ts]: I'm a bit confused about this code. Isn't enough to do: 

```ts
threadId: this.threadId,
shortenedUrl: this.shortenedUrl,
```

**Review by genaris — APPROVED** [review]
Just a minor comment regarding the code in tags, otherwise LGTM!

File: packages/shorten-url/src/repository/DidCommShortenUrlRecord.ts (hunk 7 above)
Comment (gabrielmatau79) [reply packages/shorten-url/src/repository/DidCommShortenUrlRecord.ts]: Done in the commit [d5e73d6](https://github.com/2060-io/credo-ts-didcomm-ext/pull/109/commits/d5e73d6ff141f246c08d6d2fa2e3834b8074a4ad)
